expyfun/_sound_controllers/_sounddevice.py

```diff
@@ -0,0 +1,205 @@
+        self.clear()
+
+    def _get_stream(self, fs, n_channels, api, name, api_options):
+        """Select the API and device."""
+        # API
+        if api is None:
+            api = get_config("SOUND_CARD_API", None)
+        if api is None:
+            # Eventually we should maybe allow 'Windows WDM-KS',
+            # 'Windows DirectSound', or 'MME'
+            api = dict(
+                darwin="Core Audio",
+                win32="Windows WASAPI",
+                linux="ALSA",
+                linux2="ALSA",
+            )[sys.platform]
+        key = (fs, n_channels, api, name)
+        if key not in self:
+            self[key] = _init_stream(fs, n_channels, api, name, api_options)
+        return self[key]
+
+
+_stream_registry = _StreamRegistry()
+
+
+def _init_stream(fs, n_channels, api, name, api_options=None):
+    devices = sd.query_devices()
+    if len(devices) == 0:
+        raise OSError("No sound devices found!")
+    apis = sd.query_hostapis()
+    valid_apis = []
+    for ai, this_api in enumerate(apis):
+        if this_api["name"] == api:
+            api = this_api
+            break
+        else:
+            valid_apis.append(this_api["name"])
+    else:
+        m = 'Could not find host API %s. Valid choices are "%s"'
+        raise RuntimeError(m % (api, ", ".join(valid_apis)))
+    del this_api
+
+    # Name
+    if name is None:
+        name = get_config("SOUND_CARD_NAME", None)
+    if name is None:
+        global _DEFAULT_NAME
+        if _DEFAULT_NAME is None:
+            di = api["default_output_device"]
+            _DEFAULT_NAME = devices[di]["name"]
+            logger.exp("Selected default sound device: %r" % (_DEFAULT_NAME,))
+        name = _DEFAULT_NAME
+    possible = list()
+    for di, device in enumerate(devices):
+        if device["hostapi"] == ai:
+            possible.append(device["name"])
+            if name in device["name"]:
+                break
+    else:
+        raise RuntimeError(
+            "Could not find device on API %r with name "
+            "containing %r, found:\n%s" % (api["name"], name, "\n".join(possible))
+        )
+    param_str = "sound card %r (devices[%d]) via %r" % (device["name"], di, api["name"])
+    extra_settings = None
+    if api_options is not None:
+        if api["name"] == "Windows WASAPI":
+            # exclusive mode is needed for zero jitter on Windows in testing
+            extra_settings = sd.WasapiSettings(**api_options)
+        else:
+            raise ValueError(
+                'api_options only supported for "Windows WASAPI" backend, '
+                "using %s backend got api_options=%s" % (api["name"], api_options)
+            )
+        param_str += " with options %s" % (api_options,)
+    param_str += ", %d channels" % (n_channels,)
+    if fs is not None:
+        param_str += " @ %d Hz" % (fs,)
+    try:
+        stream = sd.OutputStream(
+            samplerate=fs,
+            blocksize=fs // 5,
```

Comment:
I've removed this line in my most recent changes, I think that should allow fs to be a float without causing errors. Looks like the push hasn't updated the code here yet, not sure why since I pushed it yesterday. 
`Blocksize` should default to 0, which the sounddevice docs seem to recommend anyways. I had copied this from the code we used years ago in the lab, which I didn't write so I'm not sure why it was specified there.

Reply:
I don't fully understand why, but it seems like this was important for getting gapless playback. I can set `blocksize=int(fs/5)`, but I don't know what effect that will have if `fs` isn't divisible by 5. All the options on my computer are divisible by 5, so I'm not sure how to test it either.